### PR TITLE
Add OnPayment in OracleContract

### PR DESCRIPTION
**Pull request by Tommo-L**
Close https://github.com/neo-project/neo/issues/2093

### CODE REVIEW

**Comment by Tommo-L** [comment]
#2095 is better

File: src/neo/SmartContract/Native/Oracle/OracleContract.cs [hunk 1]
```diff
@@ -135,6 +135,13 @@ internal override void Initialize(ApplicationEngine engine)
             engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_RequestId), new StorageItem(BitConverter.GetBytes(0ul)));
         }
 
+        [ContractMethod(0_00100000, CallFlags.None)]
+        private void OnPayment(ApplicationEngine engine, UInt160 from, BigInteger amount)
+        {
+            if (engine.CallingScriptHash != Hash)
```
Comment (roman-khimov) [inline src/neo/SmartContract/Native/Oracle/OracleContract.cs:141]: So, it's like "Oracle" tokens? Maybe we need to fix GAS calls to go through `CallFromNativeContract`? Although it's `Mint` being called here...

Comment (Tommo-L) [reply src/neo/SmartContract/Native/Oracle/OracleContract.cs:141]: `GAS.Mint` cann't be called by contract, so we cann't call it through `CallFromNativeContract`.

Comment (shargon) [inline src/neo/SmartContract/Native/Oracle/OracleContract.cs:141]:
```suggestion
            if (engine.CallingScriptHash != GAS.Hash)
```

Comment (Tommo-L) [reply src/neo/SmartContract/Native/Oracle/OracleContract.cs]: No, `CallingScriptHash ` is Oracle.Hash

**Review by shargon — APPROVED** [review]

File: src/neo/SmartContract/Native/Oracle/OracleContract.cs [hunk 2]
```diff
@@ -135,6 +135,13 @@ internal override void Initialize(ApplicationEngine engine)
             engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_RequestId), new StorageItem(BitConverter.GetBytes(0ul)));
         }
 
+        [ContractMethod(0_00100000, CallFlags.None)]
+        private void OnPayment(ApplicationEngine engine, UInt160 from, BigInteger amount)
+        {
+            if (engine.CallingScriptHash != GAS.Hash)
```
Comment (erikzhang) [inline src/neo/SmartContract/Native/Oracle/OracleContract.cs:141]: The calling scripthash couldn't be `GAS.Hash`.

**Review by erikzhang — CHANGES_REQUESTED** [review]

File: src/neo/SmartContract/Native/Oracle/OracleContract.cs (hunk 2 above)
Comment (shargon) [reply src/neo/SmartContract/Native/Oracle/OracleContract.cs:141]: Change `CallingScriptHash` after call `CallFromNativeContract`?

Comment (shargon) [reply src/neo/SmartContract/Native/Oracle/OracleContract.cs:141]:
```csharp
engine.CurrentContext.GetState<ExecutionContextState>().CallingScriptHash = GAS.Hash;
```

Comment (shargon) [reply src/neo/SmartContract/Native/Oracle/OracleContract.cs:141]: Or load a GAS context before call it

Comment (erikzhang) [reply src/neo/SmartContract/Native/Oracle/OracleContract.cs:141]: Maybe we add a `MintWithoutCallingOnPayment()` method in GAS.

Comment (Tommo-L) [reply src/neo/SmartContract/Native/Oracle/OracleContract.cs:141]: Just remove `PostTransfer` in Mint and Burn method?

Comment (shargon) [reply src/neo/SmartContract/Native/Oracle/OracleContract.cs:141]: take a look to https://github.com/Tommo-L/neo/pull/21

Comment (Tommo-L) [reply src/neo/SmartContract/Native/Oracle/OracleContract.cs:141]: But it breaks the context, a little....

Comment (erikzhang) [reply src/neo/SmartContract/Native/Oracle/OracleContract.cs:141]: > Just remove PostTransfer in Mint and Burn method?

But you need to fire the `Transfer` event.

Comment (shargon) [reply src/neo/SmartContract/Native/Oracle/OracleContract.cs:141]: Please take a look tohttps://github.com/neo-project/neo/pull/2098

**Pull request closed without merging**